File: 883-car-fleet/car-fleet.java
class Solution {
    public int carFleet(int tar, int[] po, int[] sp) {
        if(tar == 0 )
            return 0;
        int cf = 1;
        Map<Integer,Integer> map = new HashMap<>();
        for(int i = 0 ;  i < po.length; i++){
            map.put(po[i], sp[i]);
        }
        Arrays.sort(po);
        double maxT = -99;
        for(int i = po.length-1; i >= 0 ; i--){
            double time = (double)( tar - po[i])/ (double)(map.get(po[i]));
            if(maxT == -99){
                maxT = time;
                continue;
            }
            if(time > maxT){
                cf++;
                maxT = time;
            }
        }
        return cf;
    }
}

File: 883-car-fleet/car-fleet.kt
class Solution {
    fun carFleet(target: Int, position: IntArray, speed: IntArray): Int {
        val n = speed.size
        if(n<=0)
            return 0
        var ans = 1
        val map = position.indices.associate{
            position[it] to speed[it]
        }
        val pos = position.sortedArrayDescending()
        var min = -1.0
        for(ele in pos){
            val speed = map[ele]?: continue 
            val time = (target - ele) / speed.toDouble()
            if(min == -1.0){
                min = time
            }else{
                if(time > min){
                    min = time
                    ans++
                }
            }
        }
        return ans
    }
}